### PR TITLE
Nix 2.24 upgrade

### PR DESCRIPTION
**Pull request by Mic92**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/tools/package-management/nix/default.nix [hunk 1]
```diff
@@ -147,8 +147,9 @@ let
             grep -v 'nix-${pkg.version}$' <<< "$paths"
             echo
             echo "Fix it by running in nixpkgs:"
-            echo
             echo "curl https://releases.nixos.org/nix/nix-${pkg.version}/fallback-paths.nix >nixos/modules/installer/tools/nix-fallback-paths.nix"
+            echo "Or go to the current maintainance branch on https://hydra.nixos.org/project/nix to get the latest eval of nix and run:"
+            echo "nixos/modules/installer/tools/update-fallback-paths.py <hydra-eval-id>"
```
Comment (Mic92) [inline pkgs/tools/package-management/nix/default.nix:152]: I added this as a learning from the post-mortem in the nix bump: https://discourse.nixos.org/t/nar-unpacking-vulnerability-post-mortem/52301

**Review by fabianhjr — APPROVED** [review]
Release notes didn't get reverted and are still present on rl-2411

**Comment by fabianhjr** [comment]
Mic92, just an idea (not a request / completely to your consideration),

Maybe target the current staging-next cycle to test before a broader rollout? (It isn't many rebuilds and could go directly to the main branch but due to the criticallity of nix for NixOS might be preferable to reduce the blast radius a bit)

**Comment by Mic92** [comment]
> Mic92, just an idea (not a request / completely to your consideration),
> 
> Maybe target the current staging-next cycle to test before a broader rollout? (It isn't many rebuilds and could go directly to the main branch but due to the criticallity of nix for NixOS might be preferable to reduce the blast radius a bit)

Sounds good. Also we can than fix instead of revert.

**Comment by Mic92** [comment]
I found some issue with the nix-upgrade service. We need to patch switch-to-configuration to ignore /nix/store transient units.

**Comment by Mic92** [comment]
> I found some issue with the nix-upgrade test. We need to patch switch-to-configuration to ignore /nix/store transient units.

Also I think this is issue would also happen with the current nix version as systemd creates a nix-store.mount unit that it than tries to unmount.

**Comment by Mic92** [comment]
So the nix-update tests revealed a regression in switch-to-configuration-ng, which needs to be fixed nevertheless as it will break any nix update: https://github.com/NixOS/nixpkgs/pull/343709

**Comment by Mic92** [comment]
Result of `nixpkgs-review pr 343655` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>nix-plugin-pijul</li>
    <li>nix-plugins</li>
  </ul>
</details>
<details>
  <summary>93 packages built:</summary>
  <ul>
    <li>appvm</li>
    <li>attic-client</li>
    <li>attic-server</li>
    <li>bundix</li>
    <li>cabal2nix</li>
    <li>cached-nix-shell</li>
    <li>common-updater-scripts</li>
    <li>crate2nix</li>
    <li>crystal2nix</li>
    <li>disko</li>
    <li>dub-to-nix</li>
    <li>dydisnix</li>
    <li>fusionInventory</li>
    <li>libnixxml</li>
    <li>lua51Packages.luarocks-nix</li>
    <li>lua52Packages.luarocks-nix</li>
    <li>lua53Packages.luarocks-nix</li>
    <li>lua54Packages.luarocks-nix</li>
    <li>luajitPackages.luarocks-nix</li>
    <li>luarocks-packages-updater</li>
    <li>nim_lk</li>
    <li>niv (niv.bin ,niv.data)</li>
    <li>nix</li>
    <li>nix-binary-cache</li>
    <li>nix-bundle</li>
    <li>nix-direnv</li>
    <li>nix-du</li>
    <li>nix-index</li>
    <li>nix-init</li>
    <li>nix-pin</li>
    <li>nix-prefetch</li>
    <li>nix-prefetch-bzr</li>
    <li>nix-prefetch-cvs</li>
    <li>nix-prefetch-docker</li>
    <li>nix-prefetch-git</li>
    <li>nix-prefetch-hg</li>
    <li>nix-prefetch-scripts</li>
    <li>nix-prefetch-svn</li>
    <li>nix-required-mounts</li>
    <li>nix-required-mounts.dist</li>
    <li>nix-template</li>
    <li>nix-update</li>
    <li>nix-update-source</li>
    <li>nix-update-source.dist</li>
    <li>nix-update.dist</li>
    <li>nix-visualize</li>
    <li>nix-visualize.dist</li>
    <li>nix.debug</li>
    <li>nix.dev</li>
    <li>nix.doc</li>
    <li>nix.man</li>
    <li>nixStatic</li>
    <li>nixStatic.dev</li>
    <li>nixStatic.doc</li>
    <li>nixStatic.man</li>
    <li>nixci</li>
    <li>nixos-anywhere</li>
    <li>nixos-generators</li>
    <li>nixos-install</li>
    <li>nixos-rebuild</li>
    <li>nixos-shell</li>
    <li>nixpkgs-hammering</li>
    <li>nixpkgs-manual</li>
    <li>nixpkgs-review</li>
    <li>nixpkgs-review.dist</li>
    <li>nixtract</li>
    <li>node2nix</li>
    <li>npins</li>
    <li>nuget-to-nix</li>
    <li>nurl</li>
    <li>nvfetcher</li>
    <li>outline</li>
    <li>prefetch-yarn-deps</li>
    <li>python311Packages.nix-kernel</li>
    <li>python311Packages.nix-kernel.dist</li>
    <li>python312Packages.nix-kernel</li>
    <li>python312Packages.nix-kernel.dist</li>
    <li>sbomnix</li>
    <li>sbomnix.dist</li>
    <li>sonarr</li>
    <li>swiftPackages.swiftpm2nix</li>
    <li>terranix</li>
    <li>typescript-language-server</li>
    <li>update-nix-fetchgit</li>
    <li>update-python-libraries</li>
    <li>vimPluginsUpdater</li>
    <li>vulnix</li>
    <li>vulnix.dist</li>
    <li>vulnix.doc</li>
    <li>vulnix.man</li>
    <li>wp4nix</li>
    <li>yarn2nix</li>
    <li>zon2nix</li>
  </ul>
</details>

File: pkgs/by-name/ni/nix-plugin-pijul/package.nix [hunk 2]
```diff
@@ -31,7 +30,7 @@ stdenv.mkDerivation (finalAttrs: {
   buildInputs = [
     boost
     howard-hinnant-date
-    nix
+    nixVersions.nix_2_18
```
Comment (Mic92) [inline pkgs/by-name/ni/nix-plugin-pijul/package.nix:33]: cc @2xsaiko

Comment (Mic92) [reply pkgs/by-name/ni/nix-plugin-pijul/package.nix:33]: Doesn't build with nix 2.24.

**Comment by Mic92** [comment]
@vcunat gave me green light to merge this to staging-next.

**Comment by trofi** [comment]
829636f4173acb044ee430037762e2b30a03cb1e broke the test eval in `staging-next` as:

```
$ nix build --no-link -f. nix-plugin-pijul.tests
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:34:12:
           33|
           34|   strict = derivationStrict drvAttrs;
             |            ^
           35|

       … while evaluating derivation 'localRepoCheck-nix-2.25.0pre20240920_ca3fc169'
         whose name attribute is located at pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'buildCommand' of derivation 'localRepoCheck-nix-2.25.0pre20240920_ca3fc169'
         at pkgs/build-support/trivial-builders/default.nix:59:17:
           58|         enableParallelBuilding = true;
           59|         inherit buildCommand name;
             |                 ^
           60|         passAsFile = [ "buildCommand" ]

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: function 'anonymous lambda' called with unexpected argument 'nix'
       at pkgs/by-name/ni/nix-plugin-pijul/package.nix:1:1:
            1| {
             | ^
            2|   lib,
       Did you mean lib?
```

**Comment by trofi** [comment]
Proposed a simple fix as:
- https://github.com/NixOS/nixpkgs/pull/346689

**Pull request merged**